src/models/EarthEngine.ts

```diff
@@ -0,0 +1,123 @@
+import _ from "lodash";
+import moment, { Moment } from "moment";
+import ee from "@google/earthengine";
+import { GeometryPoint, GeometryPolygon, ImageCollection } from "@google/earthengine";
+import { InfoDataRowBase, InfoData } from "@google/earthengine";
+
+export type DataSetId = string;
+export type Coordinates = [number, number];
+export type Interval = { type: "daily"; start: Moment; end: Moment };
+
+export type Geometry =
+    | { type: "point"; coordinates: Coordinates }
+    | { type: "multi-polygon"; polygonCoordinates: Coordinates[][][] };
+
+export interface GetDataOptions<Band> {
+    id: DataSetId;
+    bands: Band[];
+    geometry: Geometry;
+    interval: Interval;
+    scale?: number;
+}
+
+export type DataItem<Band> = {
+    periodId: string;
+    date: Moment;
+    lat: number;
+    lon: number;
+    band: Band;
+    value: number;
+};
+
+export type GeeData<Band> = DataItem<Band>[];
+
+export interface Credentials {
+    client_id: string;
+    access_token: string;
+    expires_in: number;
+}
+
+type GeeGeometry = GeometryPoint | GeometryPolygon;
+
+export class EarthEngine {
+    static async init(credentials: Credentials) {
+        ee.data.setAuthToken(
+            credentials.client_id,
+            "Bearer",
+            credentials.access_token,
+            credentials.expires_in,
+            null,
+            null,
+            false
+        );
+
+        await new Promise((resolve, reject) => {
+            ee.initialize(null, null, resolve, reject);
+        });
+
+        return new EarthEngine();
+    }
+
+    async getData<Band extends string>(options: GetDataOptions<Band>): Promise<GeeData<Band>> {
+        const { id, bands, geometry, interval, scale = 30 } = options;
+        const imageCollection = new ee.ImageCollection(id);
+        const startDate = getDayString(interval.start);
+        const endDate = getDayString(interval.end); // last day is not included
+        const engineGeometry = getGeometry(geometry);
+        console.log("ee.getImageCollection", { id, bands, startDate, endDate, geometry, scale });
+
+        const rows = await getInfo(
+            imageCollection
+                .select(bands)
+                .filterDate(startDate, endDate)
+                .getRegion(engineGeometry, scale)
+        );
+
+        const header = rows[0];
```

Comment:
This access can throw a _(TypeError): Cannot read property '0' of undefined_ in case the getInfo function returns undefined. In my opinion we should either handle this in the EarthEngine.getInfo() function or throw an error in case 0 is not in rows.

Reply:
Good point. `ee.getInfo` has only one callback used both to signal data and error, now I use the 2nd argument (error) to reject the promise if needed: 1b7cec2a